File: python3/Heaps/median_of_an_integer_stream.py
import heapq


class MedianOfAnIntegerStream:
    def __init__(self):
        self.left_half = []  # Max-heap.
        self.right_half = []  # Min-heap.

    def add(self, num: int) -> None:
        # If 'num' is less than or equal to the max of 'left_half', it 
        # belongs to the left half.
        if not self.left_half or num <= -self.left_half[0]:
            heapq.heappush(self.left_half, -num)
            # Rebalance the heaps if the size of the 'left_half' 
            # exceeds the size of the 'right_half' by more than one.
            if len(self.left_half) > len(self.right_half) + 1:
                heapq.heappush(self.right_half, -heapq.heappop(self.left_half))
        # Otherwise, it belongs to the right half.
        else:
            heapq.heappush(self.right_half, num)
            # Rebalance the heaps if 'right_half' is larger than 
            # 'left_half'.
            if len(self.left_half) < len(self.right_half):
                heapq.heappush(self.left_half, -heapq.heappop(self.right_half))

    def get_median(self) -> float:
        if len(self.left_half) == len(self.right_half):
            return (-self.left_half[0] + self.right_half[0]) / 2.0
        return -self.left_half[0]

File: csharp/Heaps/MedianOfAnIntegerStream.cs
public class MedianOfAnIntegerStream
{
    PriorityQueue<int, int> leftHalf;   // max-heap
    PriorityQueue<int, int> rightHalf;  // min-heap

    public MedianOfAnIntegerStream()
    {
        leftHalf = new(Comparer<int>.Create((x, y) => y - x));
        rightHalf = new(Comparer<int>.Create((x, y) => x - y));
    }

    public void Add(int num)
    {
        // If 'num' is less than or equal to the max of 'leftHalf', it 
        // belongs to the left half.
        if (leftHalf.Count == 0 || num <= leftHalf.Peek())
        {
            leftHalf.Enqueue(num, num);

            // Rebalance the heaps if the size of the 'leftHalf' 
            // exceeds the size of the 'rightHalf' by more than one.
            if (leftHalf.Count > rightHalf.Count + 1)
            {
                int leftTop = leftHalf.Dequeue();
                rightHalf.Enqueue(leftTop, leftTop);
            }
        }
        // Otherwise, it belongs to the right half.
        else
        {
            rightHalf.Enqueue(num, num);

            // Rebalance the heaps if 'rightHalf' is larger than 'leftHalf'.
            if (leftHalf.Count < rightHalf.Count)
            {
                int rightTop = rightHalf.Dequeue();
                leftHalf.Enqueue(rightTop, rightTop);
            }
        }
    }

    public double GetMedian()
    {
        if (leftHalf.Count == rightHalf.Count)
            return (leftHalf.Peek() + rightHalf.Peek()) / 2.0;

        return leftHalf.Peek();
    }
}
